---
title: "3_alpha_diversity"
author: "Pedro Beschoren da Costa"
date: "August 27, 2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


# Loading packages

# load libraries and data
```{r}

#Set working directory to project directory if starting fresh
#setwd("./")
#getwd() #ok

#Handling data 
library("dplyr")
library("tibble")
library("stringr")  # to wrangle string vectors
library("tidyr")
library("ggrepel")  # to avoid legends overlapping in your plot
library("ggpubr")
library("purrr")  # has map() to select table elements

#Microbiome
library("vegan")  # for several essential statistical tests
library("forcats")
library("metagMisc")  # lets you create lists of split phyloseq objects
library("phyloseq")
library("decontam")
library("metagenomeSeq")
library("Biostrings")
library("mikropml")

#Visuals
library("ggplot2")

#stats
library("agricolae")  # includes some ANOVA post-hoc options
library("minpack.lm")  # lets you do some HSD tests, output is a nice table
library("car")  # for Levene's test

```

Load environment from 2_beta_diversity.Rmd script if starting this script from scratch
```{r}
#load("./R output/env.2_beta_diversity.RData")

#let's clean what its not necessary 
rm(a, AT_Endo, AT_Rhizo, BO_Endo, BO_Rhizo, b, centroids, df_info, df_NMDS, df_NMDS_split, df_NMDS_split_full,
   df_NMDS_split2, df_summ, df_UMAP,p, perm_by4, perm2, nmds_rootAndSoil, NMDS_4_slice, dip_result3,
   three_way_permanova, s.bray, x,y, sample_data, bray, contaminants)

```


## Settings for plots 
```{r}

#Set plots theme
ggplot2::theme_set(theme_bw())

# Set default axis looks
axis_looks <- theme(axis.text.x = element_text(colour = "black", size = 9,
                                               face = "bold", angle=0, hjust=0.5))+ #text in X
  theme(axis.text.y = element_text(colour = "black", size = 9, face = "bold"))+ #text in y
  theme(axis.title=element_text(size=11, face = "bold"))+ #all texts
  theme(axis.title.y = element_text(margin = ggplot2::margin(t = 0, r = 10, b = 0, l = 0)))+
  theme(panel.grid.major = element_blank(), #no grid
        panel.grid.minor = element_blank()) #no grid

#Set colors
color_treatments <- c("#FBF2C4", "#DEA868", "#C7522B", "#9BBAA0") #"Control", "MeJA 0.1mM", "MeJA 1mM", "P. brassicae OS"
color_species <- c("#146627ff", "#1fb695ff") #at & bo, respectively
color_compartment <- c("#e7b083ff", "#905423ff") #endo & rhizo, respectively

```



# 3.0 Basic alpha diversity analysis for pilot data
Aplha diversity will show you the quantifiable diversity within a sample - this way you compare which treatment is more or less diverse or species-rich than the other. Like in all the other steps we are evaluating here, there are countless variations and methods you could use. this is just a convenient template for you to get started.

Alpha diversity requires counts from samples with equal samples sizes. this means we cannot use the CSS transformation from the metagenomeseq package. also, heavy filtering of your data may skew some of the diversity metrics (due to the lack of rare species), so we should not over-filter the data before calculating alpha diversity.

## 3.1 simple Alpha diversity plots
```{r}

# as noted above, we will use rarefied data, filtered at at least 8 occurences in the dataset
# This is a very simple plot, but will serve for a visualization
plot_richness(physeq_filtered_rarefied, x = "MeJA_treatment", measures = c("Observed"), color = "Plant_species") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(color = "Plant species") +
  ylab("Observed number of ASVs") +
  facet_wrap(~Sample_type)

# as we could already expect, soils are more diverse than roots. for Brassica However the difference, if significant, is very small
```
## 3.1 Alpha diversity testing

Here we check shanon diversity indexes
```{r}
# Calculate richness for root and soil
total_diversity <- estimate_richness(physeq_filtered_rarefied) %>%
  rownames_to_column(var = "Sample")
total_diversity$Sample <- sub("X", "", total_diversity$Sample) # why the hell that X popped up? this fix it
alpha_diversity <- column_to_rownames(total_diversity, var = "Sample") #df with only alpha metrics


#### add diversity metrics to mapping file of phyloseq objects
# we do this so we can perform anovas, acess metadat, make nicer plots, etc
merg_to_ps <- sample_data(alpha_diversity) # makes the diversity calculations  sample_data for phyloseq oject...
sample_data(physeq_filtered_rarefied) <- merge_phyloseq(physeq_filtered_rarefied, merg_to_ps) # merge the new phyloseq object with the old phyloseq object
total_diversity <- as(sample_data(physeq_filtered_rarefied), "data.frame") # forces sample data of updated phyloseq object into a dataframe


# check homogeniety of variances
leveneTest((Shannon) ~ Plant_species * MeJA_treatment * Sample_type, data = total_diversity)

#Overall community
# two-way anova, shanon diversity index: 
tx <- with(total_diversity, interaction(MeJA_treatment, Plant_species, Sample_type)) # needed for tukey to test interaction
aovTukey <- aov(Shannon ~ tx, data = total_diversity) # needed for tukey to test interaction
Anova(lm((Shannon) ~ Plant_species * MeJA_treatment * Sample_type, data = total_diversity, contrasts = list(MeJA_treatment = contr.sum, Plant_species = contr.sum)), type = "2")
HSD.test(aovTukey, "tx", group = TRUE, console = TRUE) # post-hoc


# Lets see this on a simple boxplot for Shannon diversity index
plot_alpha_Shannon <- ggplot(total_diversity,
                       aes(x = MeJA_treatment, y = Shannon, fill = Plant_species)) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(y = "Shannon diversity index") +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  scale_fill_manual(values = color_species)+
  facet_wrap(Sample_type~Plant_species)

#get only those for Endo & Rhizo
endo <- total_diversity %>% filter(Sample_type == "Root") %>% droplevels()
rhizo <- total_diversity %>% filter(Sample_type == "Soil") %>% droplevels()

# two-way anova, shanon diversity index: 
tx <- with(rhizo, interaction(MeJA_treatment, Plant_species)) # needed for tukey to test interaction
aovTukey.r <- aov(Shannon ~ tx, data = rhizo) # needed for tukey to test interaction
Anova(lm((Shannon) ~ Plant_species * MeJA_treatment,
         data = endo))
HSD.test(aovTukey.r, "tx", group = TRUE, console = TRUE) # post-hoc


```


## 3.2 Alpha diversity testing on individual data parittions

Now that we see clear diversity effects, let's check in details to detect key pairwise differences wiht lists and loops

```{r}


# this will add the diversity metrics to a phyloseq object
add_diversity_to_physeq_object <- function(phyloseq_object) {

  # Calculate richness for root and soil of a sigle phyloseq object
  total_diversity <- estimate_richness(phyloseq_object) %>%
    rownames_to_column(var = "Sample")

  total_diversity$Sample <- sub("X", "", total_diversity$Sample) # why the hell that X popped up? this fix it
  total_diversity <- column_to_rownames(total_diversity, var = "Sample")


  #### add diversity metrics to mapping file of phyloseq objects
  # we do this so we can perform anovas, acess metadat, make nicer plots, etc
  merg_to_ps <- sample_data(total_diversity) # makes the diversity calculations  sample_data for phyloseq oject...
  sample_data(phyloseq_object) <- merge_phyloseq(phyloseq_object, merg_to_ps) # merge the new phyloseq object with the old phyloseq object
  diversity <- as(sample_data(phyloseq_object), "data.frame") # forces sample data of updated phyloseq object into a dataframe

  return(diversity)
}


# runs new custom function on a list of phyloseq objects, making a list of dataframes with alpha diversity indexes
diversity_list <- lapply(ps_list_rarefied, add_diversity_to_physeq_object)

#Now, plot it

#make a function
plot_alpha <- function(df){

p <- ggplot(df,
       aes(x = MeJA_treatment, y = Shannon, fill = MeJA_treatment)) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(y = "Shannon diversity index") +
  geom_jitter(data = df,
                 mapping = aes(x = MeJA_treatment, y = Shannon, fill = MeJA_treatment),
              shape = 21, size = 2, color = "black", alpha=0.8,
              position = position_jitterdodge(jitter.height = .1, jitter.width = .5))+
  scale_fill_manual(values = color_treatments)+
  scale_y_continuous(limits=c(2,6), breaks = seq(0,6, by =1))+
  axis_looks+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  theme(axis.text.x = element_blank(), axis.title.x = element_blank())

return(p)}

#apply it
p_alpha_4 <- lapply(diversity_list, function(df)plot_alpha(df))

#Make it a single figure 
ggarrange(p_alpha_4$Arabidopsis_thaliana.Root, p_alpha_4$Brassica_oleraceae.Root, 
          p_alpha_4$Arabidopsis_thaliana.Soil, p_alpha_4$Brassica_oleraceae.Soil,
          ncol=2, nrow=2, common.legend = TRUE)

#save plot
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/03_alpha_diversity",
  scale = 1.8,
  width = 1000,
  height = 900,
  units = "px",
  dpi = 300),
x=c("Shannon.svg","Shannon.png"))


# check levenes homogeniety over the list
lapply(diversity_list, function(x) {
  leveneTest((Shannon) ~ MeJA_treatment, data = x)
})


# since our initial object has been sliced into multiple parts, the anova is much simpler
# build model
aov_list <- lapply(diversity_list, function(x) {
  aov(lm((Shannon) ~ MeJA_treatment, data = x))
})

# run anova
lapply(aov_list, anova)

# check tukey
lapply(aov_list, function(x) {
  TukeyHSD(x, "MeJA_treatment")
})


```


# 3.3 - overwrite saved session
```{r}

save.image("./R output/env.3_alpha_diversity.RData")
```

Alpha Diversity analyised! proceed to scrip 4_neutral_models!


