---
title: "5_differential_abundance"
author: "Pedro"
date: "9/20/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

# load libraries and data

Skip if you started in the previous script 4b and just load the new packages needed for this script
```{r}
#Set working directory to project directory 
setwd("./")
getwd() #ok

#Handling data 
library("dplyr")
library("tibble")
library("stringr")  # to wrangle string vectors
library("tidyr")
library("ggrepel")  # to avoid legends overlapping in your plot
library("ggpubr")
library("purrr")  # has map() to select table elements

#Microbiome
library("vegan")  # for several essential statistical tests
library("forcats")
library("metagMisc")  # lets you create lists of split phyloseq objects
library("phyloseq")
library("decontam")
library("metagenomeSeq")
library("Biostrings")
library("mikropml")

#Visuals
library("ggplot2")

#stats
library("agricolae")  # includes some ANOVA post-hoc options
library("minpack.lm")  # lets you do some HSD tests, output is a nice table
library("car")  # for Levene's test

#New in this script
library(DESeq2)
library(viridis)
library(pheatmap)
library(ComplexHeatmap)

```

Load data if starting from scratch, otherwise clean environment a little
```{r}

#load("./R output/env.1_Loading_and_pre_processing.RData")

#let's clean what we dont need (if you come from script 4b)
rm(AT_asv_list, AT_asv_list.endo, BO_asv_list, BO_asv_list.endo, Neutral_plots_list_AT_Root, Neutral_plots_list_AT_Root_g,
   Neutral_plots_list_AT_soil, Neutral_plots_list_AT_soil_g, Neutral_plots_list_BO_Root, Neutral_plots_list_BO_Root_g,
   Neutral_plots_list_BO_soil, Neutral_plots_list_BO_soil_g, Venn_AT, Venn_BO, plot_alpha_BySample, export_venn, dip_result_AT,
   dip_result_BO)

rm(tx, dist.mat, predictions_listed_AT_Root, predictions_listed_AT_soil, predictions_listed_BO_Root, predictions_listed_BO_soil,
   neutral_list_AT_Root, neutral_list_AT_soil, neutral_list_BO_Root, neutral_list_BO_soil, fitstats_listed_AT_Root, fit_class_ps_list_AT_soil,
   fit_class_ps_list_BO_Root, fit_class_ps_list_BO_soil)

rm(fitstats_listed_AT_soil, fitstats_listed_BO_Root, fitstats_listed_BO_soil, BO_4_model_above, BO_4_model_asPredicted, BO_4_model_below,
   AT_Root_predicted_list, AT_soil_predicted_list, BO_Root_predicted_list, BO_soil_predicted_list, fit_class_ps_list_AT_Root, list_tibble,
   named_tibble_list_AT, named_tibble_list_BO, sample_data_df)

rm(ps_list_neutral_AT_Root, ps_list_neutral_AT_soil, ps_list_neutral_BO_Root, ps_list_neutral_BO_soil)

gc()

```

Skip this 
```{r}
#source("./Code/Functions/load_libraries.R")
#load("./R output/env.1_Loading_and_pre_processing.RData")
```


## 5.0 - Deseq2 differential abundance analysis on lists of comparisons
The use of lists in this analysis greatly reduces the amount of hard code. if running this for the first time, read carefully about the use of deseq2 for amplicon data. Each dataset may need tweaks on the selected p levels or minimal fold changes; for that you will edit build_sig_tab()


base tutorial used here: https://joey711.github.io/phyloseq-extensions/DESeq2.html
heatmap for final plot: http://rstudio-pubs-static.s3.amazonaws.com/288398_185f2889a5f641c6b9aa7b14fa15b635.html

### 5.1 - Load libraries and create deseq2 object
This will be the bulk of the calculations and can take a moment

```{r}
#ps_list_nonraref <- phyloseq_sep_variable(physeq_filtered, variable = c("Plant_species", "Sample_type")) # this makes one list wth 4 NON-RAREFIED objects
#[MA 28-02-25; for now the function phyloseq_sep_variable doesnt work when given 2 variables,
#instead use the `ps_root_soil` object that is already the physeq_filtered splited by 4: Script 2_beta_diversity, line 108]

ps_root_soil 
ps_root_soil <- lapply(ps_root_soil, function(ps){ prune_taxa(taxa_sums(ps) > 0, ps)}) #just checking if all aASVs with 0 reads were removed, yes.

#change name so it matches the code from Pedro
ps_list_nonraref <- ps_root_soil

# make list of deseqed phyloseq objects
ps_list_deseq <- lapply(ps_list_nonraref, function(x) phyloseq_to_deseq2(x, ~MeJA_treatment))
ps_list_deseq <- lapply(ps_list_deseq, function(x) DESeq(x, test = "Wald", fitType = "parametric"))

# this unlisted, unsplit deseq2 object will be necessary to construct the heatmap (Deseq in full community)
ps_deseq <- phyloseq_to_deseq2(physeq_filtered, ~MeJA_treatment)
ps_deseq <- DESeq(ps_deseq, test = "Wald", fitType = "parametric")

# make a list of the pairwise comparisons you want to run according the treatment ;
#  positive fold change: higher in fist factor (treatment) / negative fold change: higher in second factor(control)
pairwise_comparison <- list(
  c("MeJA_treatment", "MeJA_0.1", "control"),
  c("MeJA_treatment", "MeJA_1.0", "control"),
  c("MeJA_treatment", "oral_secretion", "control")
)
```
### 5.2 - make the pairwise comparisons
These comparisons will be based on the data splits and the pairwise_comparison list you generated in the previous chunk

```{r}
# this function will build the significance table for each list, and has to be run once for each pairwise comparison (contro_VS_Meja1.0 and Control_VS_MeJA0.1)
# I don't know why it is not working for the last contrast of the list above. it could be due to different sample sizes

build_sig_tab <- function(deseq_list, contrast_cha_vector) {
  alpha <- 0.05 # defines the alpha level for p adjustment to control for false discovery rate
  min_fold <- 4 # defines minimum fold difference in counts to classify an OTU as differential


  list_res <- lapply(deseq_list, function(x) {
    results(x, contrast = contrast_cha_vector, cooksCutoff = FALSE)
  })

  list_sigtab <- lapply(list_res, function(x) {
    x[which(x$padj < alpha), ]
  }) # cuts results based on alpha for FDR

  list_sigtab <- lapply(list_sigtab, function(x) {
    x[which(abs(x$log2FoldChange) > min_fold), ]
  }) # cuts results based on minimum fold change

  list_sigtab <- lapply(list_sigtab, function(x) {
    x[order(x$padj), ]
  }) # sorts lists of OTUs based on p values of pairwise abundance differences



  # this silenced chuck allows you to limit the number of OTUs accepted as diferential in each comparision. note that if you define the minimum number as 1:40 (1 to 40) but you only hace 39 differentially abundannt OTUs the function will crash and you will not have these results. just remove the hashtags below to limt to 3 top differentially abundant OTUs

  # max_OTU_n<-1:3 # defines the maximum number of OTUs to include in each pairwise coparison
  # list_sigtab <- lapply(list_sigtab, function (x)
  # x[max_OTU_n,])  # cuts results based on maximum numer of OTUs counted as differential



  list_sigtab <- lapply(list_sigtab, function(x) {
    cbind(as(x, "data.frame"), as(tax_table(physeq_filtered)[rownames(x), ], "matrix"))
  }) # here we only colect the taxa names from the original, unsplit, phyloseq object
  return(list_sigtab)
}



# now use the custom function to build the significance tables for each comparison defined in the contrasts.
# these represent only treatment vs control comparisons

#MeJA 0.1 vs Control
Sigtab1 <- build_sig_tab(
  deseq_list = ps_list_deseq,
  contrast_cha_vector = pairwise_comparison[[1]]
)

#MeJA 1 vs Control
Sigtab2 <- build_sig_tab(
  deseq_list = ps_list_deseq,
  contrast_cha_vector = pairwise_comparison[[2]]
)

#OS vs Control
Sigtab3 <- build_sig_tab(
  deseq_list = ps_list_deseq,
  contrast_cha_vector = pairwise_comparison[[3]]
) # we can use lists here too, mayme mapply() or map() but I won't fix it at this point


# now extract the rownames of the significance tables and put it into a single vector
select_listed <- do.call(c, list(Sigtab1, Sigtab2, Sigtab3)) %>% # concatenate the lists resulting from build_sig_tab(), then...
  lapply(function(x) rownames(x)) %>% # extract the rownames from the list, then....
  unlist(use.names = FALSE) %>% # flatten the rowname (OTU ID) list, ignoring the names of the lists, then...
  unique() # remove duplications of the same OTUs
```

#Save DA output

```{r}

# Example assuming you have one significance table (adjust if you have a list of them)
ps_list_deseq
ps_deseq
pairwise_comparison
select_listed #unique ASVs 117
Sigtab1
Sigtab2
Sigtab3

#make a list 
DA_results <- list(ps_list_deseq = ps_list_deseq,
                   ps_deseq = ps_deseq, 
                   pairwise_comparison = pairwise_comparison,
                   meja0.1 = Sigtab1,
                   meja1 = Sigtab2,
                   caterpillarOS = Sigtab3,
                   select_listed = select_listed)

#save results 
saveRDS(DA_results, "./R output/05_differential_abundance/DA_results.rds")

#clean
rm(DA_results, ps_list_deseq, ps_deseq, pairwise_comparison, Sigtab1, Sigtab2, Sigtab3, select_listed)

```

#Visualization[PENDING]
```{r}

```




### 5.3 - Construct heatmaps
A heatmap like this is a convinient way to represent the key differences across OTUs and treatments

note that on this heatmap, all comparisons are made against the control. you will need to change the contrasts to make, for example, a 1.0 Meja - 0.1MeJA comparison
```{r}
# finally, build the heatmap
# note that the log change calculations are based on the full, unsplit and unlisted deseq2 object (that is, our complete dataset)
ntb <- normTransform(ps_deseq) # defaults to log2(x+1)
log2.norm.countsb <- assay(ntb)[select_listed, ] # these colnames should match rownames(df)
df <- as.data.frame(colData(ps_deseq)[, c("MeJA_treatment", "Sample_type", "Plant_species")])
row.names(df) <- row.names(as.data.frame(colData(ps_deseq)))
df_row <- as.data.frame((tax_table(physeq_filtered))[, c("Family")])
# this chnk of code lets us remove some families form the legend
df_row$main_families <- df_row$Family %in% c(
  "f__Burkholderiaceae", # define if the family is on this list of famillies...
  "f__Chitinophagaceae",
  "f__Comamonadaceae",
  "f__Flavobacteriaceae",
  "f__Oxalobacteraceae",
  "f__Rhizobiaceae",
  "f__Sphingobacteriaceae",
  "f__Streptomycetaceae",
  "f__Xanthobacteraceae"
)
df_row <- mutate(df_row, main_fam = if_else(main_families == "TRUE", df_row$Family, "Other")) %>% # make new variable: if the family is on the lsit of families, add it to the new variale, if not wirte is as "Other"
  select(main_fam) # remove old columns

# now plot the ehatmap
pheatmap::pheatmap(log2.norm.countsb,
  color = viridis(24), annotation_col = df, annotation_row = df_row, main = "log2(counts + 1)",
  scale = "none", fontsize_row = 3, fontsize_col = 8
)


# Alternative: if you are dealing with too many samples and it's difficult to read, plot one list element at a time with this function:
mapply(function(list_deseq_objects, list_ps_objects) {
  ntb <- normTransform(list_deseq_objects) # defaults to log2(x+1)
  differential_OTUs_in_deseq_list_element <- select_listed[which(select_listed %in% row.names(assay(ntb)))] # subsets selection of differentially abundandt OTUs so that only those represented in the deseq2 list portion are present
  log2.norm.countsb <- assay(ntb)[differential_OTUs_in_deseq_list_element, ] # these colnames should match rownames(df)
  df <- as.data.frame(colData(list_deseq_objects)[, c("MeJA_treatment", "Sample_type", "Plant_species")])
  row.names(df) <- row.names(as.data.frame(colData(list_deseq_objects)))
  df_row <- as.data.frame((tax_table(list_ps_objects))[, c("Class")])
  pheatmap::pheatmap(log2.norm.countsb,
    color = viridis(24), annotation_col = df, annotation_row = df_row, main = "log2(counts + 1)",
    scale = "none", fontsize_row = 3, fontsize_col = 8
  )
},
list_deseq_objects = ps_list_deseq,
list_ps_objects = ps_list_rarefied,
SIMPLIFY = FALSE
)

#clean
rm(log2.norm.countsb, ntb, df, df_row)

```


# 5.4 - overwrite saved session
```{r}

save.image("./R output/analysis_session.RData")

```

Differential abudance analysed! proceed to scrip 6_network analysis!



