---
title: "5_differential_abundance"
author: "Pedro"
date: "9/20/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

# load libraries and data

Skip if you started in the previous script 4b and just load the new packages needed for this script
```{r}
#Set working directory to project directory 
setwd("./")
getwd() #ok

#Handling data 
library("dplyr")
library("tibble")
library("stringr")  # to wrangle string vectors
library("tidyr")
library("ggrepel")  # to avoid legends overlapping in your plot
library("ggpubr")
library("purrr")  # has map() to select table elements

#Microbiome
library("vegan")  # for several essential statistical tests
library("forcats")
library("metagMisc")  # lets you create lists of split phyloseq objects
library("phyloseq")
library("decontam")
library("metagenomeSeq")
library("Biostrings")
library("mikropml")

#Visuals
library("ggplot2")

#stats
library("agricolae")  # includes some ANOVA post-hoc options
library("minpack.lm")  # lets you do some HSD tests, output is a nice table
library("car")  # for Levene's test

#New in this script
library(DESeq2)
library(viridis)
library(pheatmap)
library(ComplexHeatmap)

```

Load data if starting from scratch, otherwise clean environment a little
```{r}

#load("./R output/env.1_Loading_and_pre_processing.RData")

#let's clean what we dont need (if you come from script 4b)
rm(AT_asv_list, AT_asv_list.endo, BO_asv_list, BO_asv_list.endo, Neutral_plots_list_AT_Root, Neutral_plots_list_AT_Root_g,
   Neutral_plots_list_AT_soil, Neutral_plots_list_AT_soil_g, Neutral_plots_list_BO_Root, Neutral_plots_list_BO_Root_g,
   Neutral_plots_list_BO_soil, Neutral_plots_list_BO_soil_g, Venn_AT, Venn_BO, plot_alpha_BySample, export_venn, dip_result_AT,
   dip_result_BO)

rm(tx, dist.mat, predictions_listed_AT_Root, predictions_listed_AT_soil, predictions_listed_BO_Root, predictions_listed_BO_soil,
   neutral_list_AT_Root, neutral_list_AT_soil, neutral_list_BO_Root, neutral_list_BO_soil, fitstats_listed_AT_Root, fit_class_ps_list_AT_soil,
   fit_class_ps_list_BO_Root, fit_class_ps_list_BO_soil)

rm(fitstats_listed_AT_soil, fitstats_listed_BO_Root, fitstats_listed_BO_soil, BO_4_model_above, BO_4_model_asPredicted, BO_4_model_below,
   AT_Root_predicted_list, AT_soil_predicted_list, BO_Root_predicted_list, BO_soil_predicted_list, fit_class_ps_list_AT_Root, list_tibble,
   named_tibble_list_AT, named_tibble_list_BO, sample_data_df)

rm(ps_list_neutral_AT_Root, ps_list_neutral_AT_soil, ps_list_neutral_BO_Root, ps_list_neutral_BO_soil)

gc()

```

## 5.0 - Deseq2 differential abundance analysis on lists of comparisons
The use of lists in this analysis greatly reduces the amount of hard code. if running this for the first time, read carefully about the use of deseq2 for amplicon data. Each dataset may need tweaks on the selected p levels or minimal fold changes; for that you will edit build_sig_tab()


base tutorial used here: https://joey711.github.io/phyloseq-extensions/DESeq2.html
heatmap for final plot: http://rstudio-pubs-static.s3.amazonaws.com/288398_185f2889a5f641c6b9aa7b14fa15b635.html

### 5.1 - Load libraries and create deseq2 object
This will be the bulk of the calculations and can take a moment

```{r}
#ps_list_nonraref <- phyloseq_sep_variable(physeq_filtered, variable = c("Plant_species", "Sample_type")) # this makes one list wth 4 NON-RAREFIED objects
#[MA 28-02-25; for now the function phyloseq_sep_variable doesnt work when given 2 variables,
#instead use the `ps_root_soil` object that is already the physeq_filtered splited by 4: Script 2_beta_diversity, line 108]

ps_root_soil 
ps_root_soil <- lapply(ps_root_soil, function(ps){ prune_taxa(taxa_sums(ps) > 0, ps)}) #just checking if all aASVs with 0 reads were removed, yes.

#change name so it matches the code from Pedro
ps_list_nonraref <- ps_root_soil

# make list of deseqed phyloseq objects
ps_list_deseq <- lapply(ps_list_nonraref, function(x) phyloseq_to_deseq2(x, ~MeJA_treatment))
ps_list_deseq <- lapply(ps_list_deseq, function(x) DESeq(x, test = "Wald", fitType = "parametric"))

# this unlisted, unsplit deseq2 object will be necessary to construct the heatmap (Deseq in full community)
ps_deseq <- phyloseq_to_deseq2(physeq_filtered, ~MeJA_treatment)
ps_deseq <- DESeq(ps_deseq, test = "Wald", fitType = "parametric")

# make a list of the pairwise comparisons you want to run according the treatment ;
#  positive fold change: higher in fist factor (treatment) / negative fold change: higher in second factor(control)
pairwise_comparison <- list(
  c("MeJA_treatment", "MeJA_0.1", "control"),
  c("MeJA_treatment", "MeJA_1.0", "control"),
  c("MeJA_treatment", "oral_secretion", "control")
)
```
### 5.2 - make the pairwise comparisons
These comparisons will be based on the data splits and the pairwise_comparison list you generated in the previous chunk

```{r}
# this function will build the significance table for each list, and has to be run once for each pairwise comparison (contro_VS_Meja1.0 and Control_VS_MeJA0.1)
# I don't know why it is not working for the last contrast of the list above. it could be due to different sample sizes

build_sig_tab <- function(deseq_list, contrast_cha_vector) {
  alpha <- 0.05 # defines the alpha level for p adjustment to control for false discovery rate
  min_fold <- 4 # defines minimum fold difference in counts to classify an OTU as differential


  list_res <- lapply(deseq_list, function(x) {
    results(x, contrast = contrast_cha_vector, cooksCutoff = FALSE)
  })

  list_sigtab <- lapply(list_res, function(x) {
    x[which(x$padj < alpha), ]
  }) # cuts results based on alpha for FDR

  list_sigtab <- lapply(list_sigtab, function(x) {
    x[which(abs(x$log2FoldChange) > min_fold), ]
  }) # cuts results based on minimum fold change

  list_sigtab <- lapply(list_sigtab, function(x) {
    x[order(x$padj), ]
  }) # sorts lists of OTUs based on p values of pairwise abundance differences



  # this silenced chuck allows you to limit the number of OTUs accepted as diferential in each comparision. note that if you define the minimum number as 1:40 (1 to 40) but you only hace 39 differentially abundannt OTUs the function will crash and you will not have these results. just remove the hashtags below to limt to 3 top differentially abundant OTUs

  # max_OTU_n<-1:3 # defines the maximum number of OTUs to include in each pairwise coparison
  # list_sigtab <- lapply(list_sigtab, function (x)
  # x[max_OTU_n,])  # cuts results based on maximum numer of OTUs counted as differential



  list_sigtab <- lapply(list_sigtab, function(x) {
    cbind(as(x, "data.frame"), as(tax_table(physeq_filtered)[rownames(x), ], "matrix"))
  }) # here we only colect the taxa names from the original, unsplit, phyloseq object
  return(list_sigtab)
}



# now use the custom function to build the significance tables for each comparison defined in the contrasts.
# these represent only treatment vs control comparisons

#MeJA 0.1 vs Control
Sigtab1 <- build_sig_tab(
  deseq_list = ps_list_deseq,
  contrast_cha_vector = pairwise_comparison[[1]]
)

#MeJA 1 vs Control
Sigtab2 <- build_sig_tab(
  deseq_list = ps_list_deseq,
  contrast_cha_vector = pairwise_comparison[[2]]
)

#OS vs Control
Sigtab3 <- build_sig_tab(
  deseq_list = ps_list_deseq,
  contrast_cha_vector = pairwise_comparison[[3]]
) # we can use lists here too, mayme mapply() or map() but I won't fix it at this point


# now extract the rownames of the significance tables and put it into a single vector
select_listed <- do.call(c, list(Sigtab1, Sigtab2, Sigtab3)) %>% # concatenate the lists resulting from build_sig_tab(), then...
  lapply(function(x) rownames(x)) %>% # extract the rownames from the list, then....
  unlist(use.names = FALSE) %>% # flatten the rowname (OTU ID) list, ignoring the names of the lists, then...
  unique() # remove duplications of the same OTUs
```

#Save DA output

```{r}

# Example assuming you have one significance table (adjust if you have a list of them)
ps_list_deseq
ps_deseq
pairwise_comparison
select_listed #unique ASVs 117
Sigtab1
Sigtab2
Sigtab3

#make a list 
DA_results_full <- list(ps_list_deseq = ps_list_deseq,
                   ps_deseq = ps_deseq, 
                   pairwise_comparison = pairwise_comparison,
                   meja0.1 = Sigtab1,
                   meja1 = Sigtab2,
                   caterpillarOS = Sigtab3,
                   select_listed = select_listed)

#save results 
saveRDS(DA_results_full, "./R output/05_differential_abundance/DA_results_full.rds")

#clean
rm(DA_results, ps_list_deseq, ps_deseq, pairwise_comparison, Sigtab1, Sigtab2, Sigtab3, select_listed)

```

#Arrange data
```{r}
DA_results <- list(meja0.1 = DA_results_full$meja0.1,
                   meja1 = DA_results_full$meja1,
                   caterpillarOS = DA_results_full$caterpillarOS)

# Function to tidy one subcommunity result (e.g., At_Root)
tidy_da_result <- function(da_result, species, compartment, treatment) {
  da_result %>%
    rownames_to_column(var = "ASV") %>%
    mutate(Species = species,
           Compartment = compartment,
           Treatment = treatment)
}

# Combine all DA results into one dataframe
combined_da <- bind_rows(
  tidy_da_result(DA_results$meja0.1$At_Root, "A. thaliana", "Root", "meJA 0.1mM"),
  tidy_da_result(DA_results$meja0.1$At_Soil, "A. thaliana", "Soil", "meJA 0.1mM"),
  tidy_da_result(DA_results$meja0.1$Bo_Root, "B. oleracea", "Root", "meJA 0.1mM"),
  tidy_da_result(DA_results$meja0.1$Bo_Soil, "B. oleracea", "Soil", "meJA 0.1mM"),
  
  tidy_da_result(DA_results$meja1$At_Root, "A. thaliana", "Root", "meJA 1mM"),
  tidy_da_result(DA_results$meja1$At_Soil, "A. thaliana", "Soil", "meJA 1mM"),
  tidy_da_result(DA_results$meja1$Bo_Root, "B. oleracea", "Root", "meJA 1mM"),
  tidy_da_result(DA_results$meja1$Bo_Soil, "B. oleracea", "Soil", "meJA 1mM"),
  
  tidy_da_result(DA_results$caterpillarOS$At_Root, "A. thaliana", "Root", "Caterpillar OS"),
  tidy_da_result(DA_results$caterpillarOS$At_Soil, "A. thaliana", "Soil", "Caterpillar OS"),
  tidy_da_result(DA_results$caterpillarOS$Bo_Root, "B. oleracea", "Root", "Caterpillar OS"),
  tidy_da_result(DA_results$caterpillarOS$Bo_Soil, "B. oleracea", "Soil", "Caterpillar OS")
)

# Optional â€” remove prefixes like "p__", "f__", etc., for cleaner plot labels
clean_taxonomy <- function(x) gsub("^\\w__","",x)
combined_da <- combined_da %>%
  mutate(Phylum = clean_taxonomy(Phylum),
         Family = clean_taxonomy(Family),
         Genus = clean_taxonomy(Genus))

# Set ASV order (optional) to have most important ASVs on top
asv_order <- combined_da %>%
  group_by(ASV) %>%
  summarise(max_abs_lfc = max(abs(log2FoldChange))) %>%
  arrange(-max_abs_lfc) %>%
  pull(ASV)

combined_da$ASV <- factor(combined_da$ASV, levels = asv_order)

#make new label with ASV and family info
combined_da <- combined_da %>%
  mutate(ASV_Label = paste0(ASV, " (", Family, ")"))

combined_da$ASV_Label <- factor(combined_da$ASV_Label, levels = unique(combined_da$ASV_Label[order(match(combined_da$ASV, asv_order))]))

#add the combination of all of the treatments 
combined_da <- combined_da %>%
  mutate(Condition = paste(Species, Compartment, Treatment, sep = "-"))

combined_da_original <- combined_da

library(glue)
#pimp the genus names 
combined_da <- combined_da_original %>%
  mutate(Genus = ifelse(Genus == "Burkholderia-Caballeronia-Paraburkholderia", 
                        "Burkholderia Complex", 
                        Genus)) %>%
  mutate(Label_full = ifelse(is.na(Genus) | Genus == "", 
                             glue("{Family} | {ASV}"),  # Case when genus is missing
                             glue("{Family} | {Genus}_{ASV}")))  # Desired format with _

#lets fix the last "_"
combined_da <- combined_da %>% mutate(Label_full = gsub("ASV_", "ASV-", Label_full)) 

#check
head(combined_da) #ok 

#check
combined_da$Condition #ok


#plot it
ggplot(combined_da, aes(x = Treatment, y = ASV_Label, fill = log2FoldChange)) +
  geom_tile() +
  #facet_grid(Species ~ Compartment) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red") +
  labs(x = "Treatment", y = "ASV (Family)", fill = "Log2 Fold Change") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 6))

#works nice :) but I would like to have one column for each condition, first arrange data 

unique(combined_da$Family) 

#re-order some factors
library(forcats)

# Step 1: Count ASVs per Phylum
phylum_order <- combined_da %>%
  dplyr::count(Phylum) %>%
  arrange(desc(n)) %>%
  pull(Phylum)

# Step 2: Order Phylum by number of ASVs and Family within Phylum (alphabetical within each phylum)
combined_da <- combined_da %>%
  mutate(Phylum = factor(Phylum, levels = phylum_order)) %>%
  arrange(Phylum, Family) %>%
  mutate(Family = factor(Family, levels = unique(Family)))  # Preserve the order in Family

combined_da  <- combined_da  %>% mutate(Treatment= factor(Treatment,
                                 levels = c("meJA 0.1mM", "meJA 1mM", "Caterpillar OS")))

#let's save it 
saveRDS(combined_da, "./R output/05_differential_abundance/combined_da_df.rds")

```


#Log2fold change plot

##Endo
```{r}
color_JAtreatments <- c( "#DEA868", "#C7522B", "#9BBAA0")
names(color_JAtreatments) <- c("meJA 0.1mM", "meJA 1mM", "Caterpillar OS")

#plot it
plot_Endo <-ggplot(subset(combined_da, Compartment == "Root"),
       aes(x = Family, y = log2FoldChange, color = Treatment)) +
  geom_point(size = 4, alpha = 0.5)+
  labs(y = "log2(Fold Change)", color = "Treatment") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))+
  scale_color_manual(values=color_JAtreatments)+
  geom_hline(yintercept = 0, linetype = "dashed", color = "black")+
  facet_grid(Species~Phylum, scales = "free_x", space = "free_x")+
  theme(axis.text.x = element_text(colour = "black", size = 10,
                                   face = "bold", angle=90, hjust=1))

plot_Endo

mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/05_differential_abundance",
  scale = 1.8,
  width = 2000,
  height = 1000,
  units = "px",
  dpi = 300),
x=c("log2_Endo.svg","log2_Endo.png"))

```

##Rhizo
```{r}
plot_Rhizo <-ggplot(subset(combined_da, Compartment == "Soil"),
       aes(x = Family, y = log2FoldChange, color = Treatment)) +
  geom_point(size = 4, alpha = 0.5)+
  labs(y = "log2(Fold Change)", color = "Treatment") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))+
  scale_color_manual(values=color_JAtreatments)+
  geom_hline(yintercept = 0, linetype = "dashed", color = "black")+
  facet_grid(Species~Phylum, scales = "free_x", space = "free_x")

plot_Rhizo 

mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/05_differential_abundance",
  scale = 1.5,
  width = 1500,
  height = 800,
  units = "px",
  dpi = 300),
x=c("log2_Rhizo.svg","log2_Rhizo.png"))

```

##Full
```{r}

full <- ggplot(combined_da,
       aes(x = Family, y = log2FoldChange, color = Treatment)) +
  geom_point(size = 4, alpha = 0.6) +
  labs(y = "log2(Fold Change)", color = "Treatment") +
  scale_color_manual(values = color_JAtreatments) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "black") +
  facet_grid(Compartment + Species ~ Phylum, scales = "free_x", space = "free_x") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 9, face = "bold"))+
  theme(legend.position = "top")

mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/05_differential_abundance",
  scale = 1.4,
  width = 2000,
  height = 1500,
  units = "px",
  dpi = 300),
x=c("log2_Full.svg","log2_Full.png"))



```

#Pheatmaps [Not used]
##Endo
Let's do one per compartment & use pheatmap
```{r}

## Endo
endo_da <- combined_da %>%
  filter(Compartment == "Root") %>%
  droplevels() %>% 
  mutate(Condition = factor(Condition))

endo_da$Condition

# Set desired column order
order_endo <- c(
  "A. thaliana-Root-meJA 0.1mM", "A. thaliana-Root-meJA 1mM", "A. thaliana-Root-Caterpillar OS",
  "B. oleracea-Root-meJA 0.1mM", "B. oleracea-Root-meJA 1mM", "B. oleracea-Root-Caterpillar OS")

endo_da <- endo_da %>%
            mutate(Condition = factor(Condition, levels = order_endo)) #re-arrange

endo_da <- endo_da %>% arrange(Family)

#pivot to wide
heatmap_data_endo <- endo_da %>%
  select(Label_full, Condition, log2FoldChange) %>%
  pivot_wider(names_from = Condition, values_from = log2FoldChange, values_fill = 0) %>%
  column_to_rownames("Label_full")

#re-arrange columns
heatmap_data_endo <- heatmap_data_endo[,order_endo]
colnames(heatmap_data_endo) #ok

# Optional annotation (Family info)
row_anno_endo <- endo_da %>%
  select(Label_full, Family) %>%
  distinct() %>%
  column_to_rownames("Label_full")

#get vector for main families 
main_families <- c(
  "Burkholderiaceae", 
  "Chitinophagaceae",
  "Comamonadaceae",
  "Flavobacteriaceae",
  "Oxalobacteraceae",
  "Rhizobiaceae",
  "Sphingobacteriaceae",
  "Streptomycetaceae",
  "Xanthobacteraceae")

# Replace Family with "Other" if it's not in main_families
row_anno_endo <- row_anno_endo %>% 
  mutate(Family = ifelse(Family %in% main_families, Family, "Other"))

# select colors for factors
#make colors for phylum with custom function
colourCount <- length(unique(row_anno_endo$Family))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Set3"))

#apply to phylum
Family_col = getPalette(colourCount)
Family_col <- c("#8DD3C7", "#FFFFB3", "#FB8072","#BEBADA", 
                "#80B1D3", "#FDB462", "#B3DE69", "#FCCDE5",
                "#A6CEE3")

family_colors <- setNames(Family_col, main_families)
family_colors["Other"] <- "#D9D9D9"


# Plot heatmap
pheatmap.endo <- pheatmap(as.matrix(heatmap_data_endo),
         color = colorRampPalette(c("blue", "white", "red"))(30),
         cluster_cols = FALSE,
         cluster_rows = FALSE,
         show_rownames = TRUE,
         annotation_row = row_anno_endo,
         annotation_colors = list(Family=family_colors),
         #annotation_col = ,
         fontsize_row = 5,
         fontsize_col = 8,
         border_color = "gray",
         main = "Differential Abundance - Endosphere")


library(ggplotify)

gg_p <- as.ggplot(pheatmap.endo) 
gg_p

#save heatmap
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/05_differential_abundance",
  scale = 1.5,
  width = 1400,
  height = 2000,
  units = "px",
  dpi = 300),
x=c("Pheatmap_Endo.svg","Pheatmap_Endo.png"))


```

##Rhizo
```{r}
## Rhizo
rhizo_da <- combined_da %>%
  filter(Compartment == "Soil") %>% 
  droplevels() %>% 
  mutate(Condition = factor(Condition))

# Set desired column order
order_rhizo <- c(
  "A. thaliana-Soil-meJA 0.1mM", "A. thaliana-Soil-meJA 1mM", "A. thaliana-Soil-Caterpillar OS",
  "B. oleracea-Soil-meJA 0.1mM", "B. oleracea-Soil-meJA 1mM", "B. oleracea-Soil-Caterpillar OS")

rhizo_da <- rhizo_da %>%
            mutate(Condition = factor(Condition, levels = order_rhizo)) #re-arrange

rhizo_da <- rhizo_da %>% arrange(Family)

#pivot to wide
heatmap_data_rhizo <- rhizo_da %>%
  select(Label_full, Condition, log2FoldChange) %>%
  pivot_wider(names_from = Condition, values_from = log2FoldChange, values_fill = 0) %>%
  column_to_rownames("Label_full")

#re-arrange columns
heatmap_data_rhizo <- heatmap_data_rhizo[,order_rhizo]
colnames(heatmap_data_rhizo) #ok

# Optional annotation (Family info)
row_anno_rhizo <- rhizo_da %>%
  select(Label_full, Family) %>%
  distinct() %>%
  column_to_rownames("Label_full")

#get vector for main families 
main_families <-  c(
  "Burkholderiaceae", 
  "Chitinophagaceae",
  "Comamonadaceae",
  "Flavobacteriaceae",
  "Oxalobacteraceae",
  "Rhizobiaceae")

# Replace Family with "Other" if it's not in main_families
row_anno_rhizo <- row_anno_rhizo %>% 
  mutate(Family = ifelse(Family %in% main_families, Family, "Other"))

# select colors for factors
unique(row_anno_rhizo$Family)

#apply to phylum
Family_colR <- c("#8DD3C7", "#FFFFB3", "#FB8072","#BEBADA", 
                "#80B1D3", "#FDB462")

family_colorsR <- setNames(Family_colR, main_families)
family_colorsR["Other"] <- "#D9D9D9"


# Plot heatmap
pheatmap.rhizo <- pheatmap(as.matrix(heatmap_data_rhizo),
         color = colorRampPalette(c("blue", "white", "red"))(30),
         cluster_cols = FALSE,
         cluster_rows = FALSE,
         show_rownames = TRUE,
         annotation_row = row_anno_rhizo,
         annotation_colors = list(Family=family_colorsR),
         border_color="gray",
         fontsize_row = 8,
         fontsize_col = 8,
         main = "Differential Abundance - Rhizosphere")

gg_pR <- as.ggplot(pheatmap.rhizo) 
gg_pR

#save heatmap
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/05_differential_abundance",
  scale = 1.4,
  width = 1400,
  height = 2000,
  units = "px",
  dpi = 300),
x=c("Pheatmap_Rhizo.svg","Pheatmap_Rhizo.png"))


```



#Save data
```{r}

saveRDS(combined_da, "./R output/05_differential_abundance/combined_da_df.rds")

```









Differential abudance analysed! proceed to scrip 6_network analysis!



