---
title: "10_plant_traits"
author: "Marcela Aragon"
date: "8/17/2022"
editor_options: 
  chunk_output_type: console
---

# 9.0 Plant traits 

During the experiment, several plant traits were measured to characterize the plant's phenotypic response to the different stress treatments. For both *A. thaliana* and *B. oleracea* 1. Dry weight was measured. Additionally 2. Leaf area, 3. Leaf length and 4.Leaf width were measured only for *B. oleraceae* and  5.Number of fruits and 6.Inflorescence dry weight were measure only for Arabidopsis as plants were already flowering. 

Finally, as both plants are phenotypically different *per se* we compared the effects of the treatments between them using Cohen's D 

## Setting up R-studio

```{r}

# set working directory to project directory 

# this will load all key libraries for the whole project
library(tidyr)
library(dplyr)
library(vegan)
library(Hmisc)
library(tibble)
library(ggpubr)
if (!require(DHARMa)) install.packages("DHARMa")
library(DHARMa)
library(lmtest)
if (!require(fitdistrplus)) install.packages("fitdistrplus")
library(fitdistrplus)
if (!require(emmeans)) install.packages("emmeans")
library(emmeans)
if (!require(AICcmodavg)) install.packages("AICcmodavg")
library(AICcmodavg)
if (!require(jtools)) install.packages("jtools")
library(jtools)
library(car)
if (!require(multcompView)) install.packages("multcompView")
library(multcompView)
library(multcomp)
if (!require(lsr)) install.packages("lsr")
library(lsr)
if (!require(effsize)) install.packages("effsize")
library(effsize)
library(gridExtra)
#source("./renv/0_load_libraries.R")


```

## Settings for plots

```{r}

# Setting up white background for plots
theme_set(theme_bw())

# Set colors 

#colors & treatments 
#color_treatments <- c("#999999", "#E69F00", "#56B4E9", "#009E73")
#names(color_treatments) <- c("Control", "MeJA 0.1mM", "MeJA 1mM", "P. brassicae OS")

#color_treatments <- hcl.colors(5,palette="Geyser")
color_treatments <- c("#FBF2C4", "#DEA868", "#C7522B", "#9BBAA0") #"Control", "MeJA 0.1mM", "MeJA 1mM", "P. brassicae OS"
color_species <- c("#146627ff", "#1fb695ff") #at & bo, respectively

color_compartment <- c("#e7b083ff", "#905423ff") #endo & rhizo, respectively

# Set axis looks
axis_looks <- theme(axis.text.x = element_text(colour = "black", size = 9,
                                               face = "bold", angle=0, hjust=0.5))+
  theme(axis.text.y = element_text(colour = "black", size = 9, face = "bold"))+
  theme(axis.title=element_text(size=11, face = "bold"))+
  theme(axis.title.y = element_text(margin = margin(t = 0, r = 10, b = 0, l = 0)))+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())+
  theme(legend.position="none")+
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  theme(axis.line = element_line(colour = "black", 
                      size = 0.8, linetype = "solid"))


treatment_names <- c("control" = "Control", "MeJA_0.1" = "0.1 mM \n MeJA",
                      "MeJA_1.0" = "1 mM \n MeJA", "oral_secretion"="P. brassicae \n OS")

```



## Load data 

```{r}
# load raw data for traits df 
raw <- read.csv("./Data/MeJA_Plant_traits_raw.csv")
str(raw )

#removing non-necessary column 
raw  <- raw [-c(7:8,16:19)]
raw[c(1:6)] <- lapply(raw[c(1:6)], factor)

#open gene expression df 
qpcr<- read.csv("./Data/qPCR_output.csv")
qpcr[1:6] <- lapply(qpcr[1:6], as.factor)
qpcr <- qpcr %>% 
        dplyr::select(id,Block,species,treatment, MYC2_at_fold_DD, lox2_at_fold_DD)%>% #keep only these columns
        dplyr::rename(myc2 = MYC2_at_fold_DD, lox2=lox2_at_fold_DD) #change names 

#merge
traits <- left_join(raw, qpcr, by="id") %>%
          mutate(gene_data =
                 case_when(
                   is.na(myc2) ~ "no", #makes new column to select later for those samples with available gene exp data
                   TRUE ~ "yes"))      #as some samples were lost by exploiding Eppendorfs 

#clean df
traits <- traits[-c(14:16)]
colnames(traits) <- gsub(".x\\b", '', colnames(traits)) #after left_join similar columns are renamed column.x and column.y
traits[c(16)] <- lapply(traits[c(16)], factor) #make gene_data column factor
str(traits)

#summary
summary(traits)

#df for Atha
Atha <- traits %>% dplyr::filter(species == "Arabidopsis_thaliana") %>% 
                   mutate(aboveground_dw = ldry_weight + inflorescence) #make new column with overall aboveground biomass
#df for Bole
Bole <- traits %>% dplyr::filter(species == "Brassica_oleraceae")

#df list 
l_sp <- list(Atha, Bole)
names(l_sp) <- c("Arabidopsis_thaliana", "Brassica_oleraceae")

```


## Load functions
```{r}

#functions to check data distribution and run GLMs

source("./Code/Functions/data_distribution_and_GLM_functions.R")

```


## 9.1 Dry weight

### Plots

```{r}

#Boxplot Atha
p.dw.Atha <- ggplot(data = Atha,
                    mapping = aes(x = treatment, y = ldry_weight), colour= treatment)+
  geom_boxplot(mapping = aes(fill= treatment), position = position_dodge(width = 0.8), lwd=0.8)+
  geom_jitter(mapping = aes(fill = treatment),
                         shape = 21, size = 1.8, color = "black", position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="Shoot biomass(g dry weight)", x=" ") +
  labs(title = "Atha")+
  scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
                     limits = c(0, max(Atha$ldry_weight*1.1))) +
  scale_x_discrete(labels=treatment_names)+
  axis_looks
  
#Boxplot Bole
p.dw.Bole <- ggplot(data = Bole,
                    mapping = aes(x = treatment, y = ldry_weight), colour= treatment)+
  geom_boxplot(mapping = aes(fill= treatment), position = position_dodge(width = 0.8), lwd=0.8)+
  geom_jitter(mapping = aes(fill = treatment),
                         shape = 21, size = 1.8, color = "black", position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="Shoot biomass(g dry weight)", x=" ") +
  labs(title = "Bole")+
  #scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
  #                   limits = c(0, max(Bole$ldry_weight*1.1))) +
  scale_x_discrete(labels=treatment_names)+
  axis_looks

#save plot
ggsave(p.dw.Bole,
       height = 3.52, width = 3.55,
       file="./R output/Figures_for_publication/traits_BO_dry_weight.svg")

```

### Data analysis

```{r}
#Arabidopsis

#Checking data distribution
continuous_distribution_plot_table(Atha$ldry_weight) #normal 
hist(Atha$ldry_weight)

#GLM
glm.Atha <- glm_gaussian_Withdf(l_sp$Arabidopsis_thaliana) #Full, but not significant 
Anova(glm.Atha$Full)
summ(glm.Atha$Full)

df_dw_At <- summary_model(glm.Atha$Full, glm.Atha$Null)

df_dw_At <- df_dw_At %>% 
            mutate(trait = "dry_weight",
                   species = "Arabidopsis_thaliana",
                   model="GLM",
                   distribution="Gaussian")

#no effect of treatment (p=0.8855) nor block (p=0.1391)in A.thaliana biomass (p=0.34)

#Brassica 
continuous_distribution_plot_table(Bole$ldry_weight) #normal 
hist(Bole$ldry_weight)

#GLM
glm.Bole <- glm_gaussian_Withdf(l_sp$Brassica_oleraceae) #Optimal
Anova(glm.Bole$Full)
summ(glm.Bole$Full, digits = 4) #significant difference 

simulateResiduals(glm.Bole$Full, plot = T) #ok, choosing this one

#### Anova 
#LM, trying with just LM
glm.Bole2 <- lm(
  ldry_weight ~ treatment + Block, 
  data = l_sp$Brassica_oleraceae)

simulateResiduals(glm.Bole2, plot = T) #ok

anova(glm.Bole2)
summary(glm.Bole2)$r.squared

summ(glm.Bole$Optimal)
summary_model(glm.Bole$Optimal, glm.Bole$Null)

#### Anova 

#Keeping GLM
df_dw_Bo <- summary_model(glm.Bole$Full, glm.Bole$Null)

df_dw_Bo <- df_dw_Bo %>% 
            mutate(trait = "dry_weight",
                   species = "Brassica_oleraceae",
                   model="GLM",
                   distribution="Gaussian")

#Post-hoc
emm.Bole <- emmeans(ref_grid(glm.Bole$Full), pairwise ~ treatment)
plot(emm.Bole$emmeans, comparisons = TRUE)

cld <- cld(emm.Bole,
           alpha = 0.05,
           Letters = letters,
           adjust = "tukey")
```

* No effect of treatment (p=0.8855) nor block (p=0.1391)in A.thaliana biomass (LM, p=0.34)
* Significant effect of treatment (p=0.004578) and no effect of Block (p=0.065722) in B. oleracea biomass (LM, p=0.0029)

--> There's only difference between 0.1 and 1% MeJA, but not against control for *B. oleraceae*, for *A. thaliana* there's no difference in dry weight. 

## 9.2 Leaf area

### Plots

```{r}

#Boxplot Bole
p.la.Bole <- ggplot(data = Bole,
                    mapping = aes(x = treatment, y = leaf_area), colour= treatment)+
  geom_boxplot(mapping = aes(fill= treatment), position = position_dodge(width = 0.8), lwd=0.8)+
  geom_jitter(mapping = aes(fill = treatment),
                         shape = 21, size = 1.8, color = "black", position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="Leaf area(cm2)", x=" ") +
  labs(title = "Bole")+
  #scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
                     #limits = c(0, max(Bole$leaf_area*1.1))) +
  scale_x_discrete(labels=treatment_names)+
  axis_looks

#save plot
ggsave(p.la.Bole,
       height = 3.52, width = 3.55,
       file="./R output/Figures_for_publication/traits_BO_leaf_area.svg")

```

### Data analysis

```{r}

#Brassica 
continuous_distribution_plot_table(Bole$leaf_area) #normal 
hist(Bole$leaf_area)

#LM
lm.la.Bole <- lm(
  leaf_area ~ treatment + Block, 
  data = l_sp$Brassica_oleraceae)

simulateResiduals(lm.la.Bole, plot = T) #ok

Anova(lm.la.Bole)
summary(lm.la.Bole)

#Post-hoc
emm.la.Bole <- emmeans(ref_grid(lm.la.Bole), pairwise ~ treatment)

plot(emm.la.Bole$emmeans, comparisons = TRUE)

cld.la <- cld(emm.la.Bole,
           alpha = 0.05,
           Letters = letters,
           adjust = "tukey")

```

MeJA 1% had significantly lower leaf area than the rest of the treatments. 

## 9.3 Leaf lenght

### Plots

```{r}

#Boxplot Bole
p.l.Bole <- ggplot(data = l_sp$Brassica_oleraceae,
                    mapping = aes(x = treatment, y = length), colour= treatment)+
  geom_boxplot(mapping = aes(fill= treatment), position = position_dodge(width = 0.8), lwd=0.8)+
  geom_jitter(mapping = aes(fill = treatment),
                         shape = 21, size = 1.8, color = "black", position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="Leaf length (cm)", x=" ") +
  axis_looks

#save plots

```

### Data analysis

```{r}

#Brassica 

#remove NA
Bole <- Bole %>% 
        dplyr::filter(id != "31") %>% droplevels()

#check distribution
continuous_distribution_plot_table(Bole$length) #normal 
hist(Bole$length)

#LM
lm.l.Bole <- lm(
  length ~ treatment + Block, 
  data = l_sp$Brassica_oleraceae)

simulateResiduals(lm.l.Bole, plot = T) #ok

Anova(lm.l.Bole)

```

No effect of treatment nor block on leaf area for *B. oleracea*

## 9.4 Leaf width 
### Plots

```{r}

#Boxplot Bole
p.w.Bole <- ggplot(data = l_sp$Brassica_oleraceae,
                    mapping = aes(x = treatment, y = width), colour= treatment)+
  geom_boxplot(mapping = aes(fill= treatment), position = position_dodge(width = 0.8), lwd=0.8)+
  geom_jitter(mapping = aes(fill = treatment),
                         shape = 21, size = 1.8, color = "black", position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="Leaf width (cm)", x=" ") +
  axis_looks

#save plots

```

### Data analysis

```{r}

#Brassica 

#check distribution
#continuous_distribution_plot_table(Bole$width) #normal 
hist(Bole$width)

#LM
lm.w.Bole <- lm(
  width ~ treatment + Block, 
  data = l_sp$Brassica_oleraceae)

simulateResiduals(lm.w.Bole, plot = T) #ok

Anova(lm.w.Bole)
summary(lm.w.Bole)

```

No significant effect of treatment nor block on leaf width for *B. oleracea*

## 9.5 Number of Fruits (Arabidopsis)

### Plots
```{r}

#Boxplot Atha
p.nf.Atha <- ggplot(data = l_sp$Arabidopsis_thaliana,
                    mapping = aes(x = treatment, y =number_fruits ), colour= treatment)+
  geom_boxplot(mapping = aes(fill= treatment), position = position_dodge(width = 0.8), lwd=0.8)+
  geom_jitter(mapping = aes(fill = treatment),
                         shape = 21, size = 1.8, color = "black", position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="number of siliques/plant", x=" ") +
  labs(title = "Atha")+
  #scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
                     #limits = c(0, max(Bole$leaf_area*1.1))) +
  scale_x_discrete(labels=treatment_names)+
  axis_looks

#save plot
ggsave(p.nf.Atha,
       height = 3.52, width = 3.55,
       file="./R output/Figures_for_publication/traits_AT_number_siliques.svg")


```

### Data analysis

```{r}

#Arabidopsis 

#check distribution
continuous_distribution_plot_table(Atha$number_fruits) #normal - gamma 
hist(Atha$number_fruits)

#LM
lm.nf.Atha <- lm(
  number_fruits ~ treatment + Block, 
  data = l_sp$Arabidopsis_thaliana)

simulateResiduals(lm.nf.Atha, plot = T) #not ok

#GLM with Gamma distribution
glm.nf.Atha <- glm (number_fruits ~ treatment + Block,
               family = Gamma (link = "log"),
               data = l_sp$Arabidopsis_thaliana)

simulateResiduals(glm.nf.Atha, plot = T) #ok

Anova(glm.nf.Atha)
summary(glm.nf.Atha)
summ(glm.nf.Atha, digits = 4)

#Post-hoc
emm.nf.Atha <- emmeans(ref_grid(glm.nf.Atha), pairwise ~ treatment, type="response")

plot(emm.nf.Atha$emmeans, comparisons = TRUE)

cld.nf <- cld(emm.nf.Atha,
           alpha = 0.05,
           Letters = letters,
           adjust = "tukey")

```

There was significantly less numbers of siliques on Arabidopsis plants treated with MeJA 1% compared to the other treatments. 


## 9.6 Inforescence dry weight (Arabidopsis)

### Plots
```{r}

#Boxplot Atha
p.inf.Atha <- ggplot(data = l_sp$Arabidopsis_thaliana,
                    mapping = aes(x = treatment, y =inflorescence ), colour= treatment)+
  geom_boxplot(mapping = aes(fill= treatment), position = position_dodge(width = 0.8), lwd=0.8)+
  geom_jitter(mapping = aes(fill = treatment),
                         shape = 21, size = 1.8, color = "black", position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="inflorescence dry weight", x=" ") +
  labs(title = "Atha")+
  #scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
                     #limits = c(0, max(Bole$leaf_area*1.1))) +
  scale_x_discrete(labels=treatment_names)+
  axis_looks

#save plot
ggsave(p.inf.Atha,
       height = 3.52, width = 3.55,
       file="./R output/Figures_for_publication/traits_AT_weight_inflorescence.svg")


```

### Data analysis

```{r}

#Arabidopsis 

#check distribution
continuous_distribution_plot_table(Atha$inflorescence) #normal - gamma 
hist(Atha$inflorescence)

#LM
lm.inf.Atha <- lm(
  inflorescence ~ treatment + Block, 
  data = l_sp$Arabidopsis_thaliana)

simulateResiduals(lm.inf.Atha, plot = T) #ok

Anova(lm.inf.Atha)
summary(lm.inf.Atha)

#Post-hoc
emm.inf.Atha <- emmeans(ref_grid(lm.inf.Atha), pairwise ~ treatment)

plot(emm.inf.Atha$emmeans, comparisons = TRUE)

cld.inf <- cld(emm.inf.Atha,
           alpha = 0.05,
           Letters = letters,
           adjust = "tukey")

```

Significantly less with MeJA 1% compared to control and MeJA 0.1%

## 9.7 Aboveground dry weight (Arabidopsis)

### Plots
```{r}

#Boxplot Atha
p.ab.Atha <- ggplot(data = l_sp$Arabidopsis_thaliana,
                    mapping = aes(x = treatment, y =aboveground_dw ), colour= treatment)+
  geom_boxplot(mapping = aes(fill= treatment), position = position_dodge(width = 0.8), lwd=0.8)+
  geom_jitter(mapping = aes(fill = treatment),
                         shape = 21, size = 1.8, color = "black", position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="aboveground dry weight", x=" ") +
  labs(title = "Atha")+
  #scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
                     #limits = c(0, max(Bole$leaf_area*1.1))) +
  scale_x_discrete(labels=treatment_names)+
  axis_looks

#save plots

```

### Data analysis

```{r}

#Arabidopsis 

#check distribution
continuous_distribution_plot_table(Atha$aboveground_dw) #normal - gamma 
hist(Atha$aboveground_dw)

#LM
lm.ab.Atha <- lm(
  aboveground_dw ~ treatment + Block, 
  data = l_sp$Arabidopsis_thaliana)

simulateResiduals(lm.ab.Atha, plot = T) #ok

Anova(lm.ab.Atha) #no effect 
summary(lm.ab.Atha)

#Post-hoc
emm.ab.Atha <- emmeans(ref_grid(lm.ab.Atha), pairwise ~ treatment)

plot(emm.ab.Atha$emmeans, comparisons = TRUE)

cld.inf <- cld(emm.ab.Atha,
           alpha = 0.05,
           Letters = letters,
           adjust = "tukey")

```

Even when using full aboveground weight (leaves+flowers) there's no effect of treatment on A. thaliana dry biomass (although now differences seem bigger)


## 9.8 LOX2 gene expression

```{r}
#make traits df longer
qpcr.long <- traits %>% 
               tidyr::pivot_longer(cols=14:15,
                     names_to="gene",
                     values_to="rel_exp")

#make summary df
qpcr.summ <- dplyr::group_by(qpcr.long, species, treatment, gene) %>%
  dplyr::summarise(
    n = dplyr::n(),
    mean = mean(rel_exp, na.rm=TRUE),
    sd = sd(rel_exp, na.rm=TRUE),
    se = sd/sqrt(n)) 

#obtain only those samples with gene expression data 
qpcr.Atha <- dplyr::filter(Atha, gene_data == "yes")
qpcr.Bole <- dplyr::filter(Bole, gene_data == "yes")

```


###Plots
```{r}

#Arabidopsis 
p.lox2.Atha <- ggplot(data = subset(qpcr.summ, species=="Arabidopsis_thaliana" & gene =="lox2"),
                    mapping = aes(x = treatment, y = mean), colour= treatment)+
  geom_bar(mapping = aes(fill = treatment), color="black", size=0.8, stat = "summary", position = position_dodge(width = 0.8))+
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.4, size=0.8, position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="relative expression of LOX2", x=" ") +
  labs(title = "Atha")+
  scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
                     limits = c(0,7)) + #y limits
  scale_x_discrete(labels=treatment_names)+
  axis_looks

#save plot
ggsave(p.lox2.Atha,
       height = 3.52, width = 3.55,
       file="./R output/Figures_for_publication/traits_AT_LOX2.svg")

#Brassica
p.lox2.Bole <- ggplot(data = subset(qpcr.summ, species=="Brassica_oleraceae" & gene =="lox2"),
                    mapping = aes(x = treatment, y = mean), colour= treatment)+
  geom_bar(mapping = aes(fill = treatment), color="black", size=0.8, stat = "summary", position = position_dodge(width = 0.8))+
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.4, size=0.8, position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="relative expression of LOX2", x=" ") +
  labs(title = "Bole")+
  scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
                     limits = c(0, 6)) + #y limits
  scale_x_discrete(labels=treatment_names)+
  axis_looks

#save plot
ggsave(p.lox2.Bole,
       height = 3.52, width = 3.55,
       file="./R output/Figures_for_publication/traits_BO_LOX2.svg")

```


###Data analysis
```{r}

#check distribution
continuous_distribution_plot_table(qpcr.Atha$lox2) #not normal, gamma 
hist(qpcr.Atha$lox2)

#LM
lm.lox2.Atha <- lm(
  lox2 ~ treatment, 
  data = qpcr.Atha)

simulateResiduals(lm.lox2.Atha, plot = T) #okay, but residuals look strange
plot(lm.lox2.Atha)

aov_residuals<-residuals(lm.lox2.Atha) 
shapiro.test(aov_residuals) #not normal 

#GLM with Gamma distribution
glm.lox2.Atha <- glm (lox2 ~ treatment,
               family = Gamma (link = "log"),
               data = qpcr.Atha)

simulateResiduals(glm.lox2.Atha, plot = T) #ok

Anova(glm.lox2.Atha)
summary(glm.lox2.Atha)
summ(glm.lox2.Atha, digits = 4)

#Post-hoc
emm.lox2.Atha <- emmeans(ref_grid(glm.lox2.Atha), pairwise ~ treatment, type="response")

plot(emm.lox2.Atha$emmeans, comparisons = TRUE)

cld.lox2.Atha <- cld(emm.lox2.Atha,
           alpha = 0.05,
           Letters = letters,
           adjust = "tukey")

#Brassica

#check distribution
continuous_distribution_plot_table(qpcr.Bole$lox2) #not normal, gamma or log-normal
hist(qpcr.Bole$lox2)

#LM
lm.lox2.Bole <- lm(
  lox2 ~ treatment, 
  data = qpcr.Atha)

simulateResiduals(lm.lox2.Bole, plot = T) 
plot(lm.lox2.Bole)

aov_residuals<-residuals(lm.lox2.Bole) 
shapiro.test(aov_residuals) #not normal 

#GLM with Gamma distribution
glm.lox2.Bole <- glm (lox2 ~ treatment,
               family = Gamma (link = "log"),
               data = qpcr.Bole)

simulateResiduals(glm.lox2.Bole, plot = T) #ok
plot(glm.lox2.Bole) #much better

Anova(glm.lox2.Bole)
summary(glm.lox2.Bole)
summ(glm.lox2.Bole, digits = 4)

#Post-hoc
emm.lox2.Bole <- emmeans(ref_grid(glm.lox2.Bole), pairwise ~ treatment, type="response")

plot(emm.lox2.Bole$emmeans, comparisons = TRUE)
pairs(emm.lox2.Bole)

cld.lox2.Bole <- cld(emm.lox2.Bole,
           alpha = 0.05,
           Letters = letters,
           adjust = "lsd")

```

In both Brassica oleracea and Arabidopsis thaliana there's a significant effect of the treatment in the relative expression levels of LOX2, specifically plants treated with the highest concentration of MeJA had higher expression of LOX2 compared to controls. 

This, however is not found back once the emmeans approach is taken. Pair contrast between Control and MeJA 1% is not significant (but barely!) p=0.0520

## 9.9 MYC2 gene expression 

### Plots
```{r}

#Arabidopsis
p.myc2.Atha <- ggplot(data = subset(qpcr.summ, species=="Arabidopsis_thaliana" & gene =="myc2"),
                    mapping = aes(x = treatment, y = mean), colour= treatment)+
  geom_bar(mapping = aes(fill = treatment), color="black", size=0.8, stat = "summary", position = position_dodge(width = 0.8))+
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.4, size=0.8, position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="relative expression of MYC2", x=" ") +
  labs(title = "Atha")+
  scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
                     limits = c(0, 12)) + 
  scale_x_discrete(labels=treatment_names)+
  axis_looks

#save plot
ggsave(p.myc2.Atha,
       height = 3.52, width = 3.55,
       file="./R output/Figures_for_publication/traits_AT_MYC2.svg")

#Brassica 
p.myc2.Bole <- ggplot(data = subset(qpcr.summ, species=="Brassica_oleraceae" & gene =="myc2"),
                    mapping = aes(x = treatment, y = mean), colour= treatment)+
  geom_bar(mapping = aes(fill = treatment), color="black", size=0.8, stat = "summary", position = position_dodge(width = 0.8))+
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.4, size=0.8, position = position_dodge(width = 0.8))+
  scale_fill_manual(values = color_treatments)+
  labs(y="relative expression of MYC2", x=" ") +
  labs(title = "Bole")+
  scale_y_continuous(expand = c(0, 0), #start graph at y = min and x = min
                     limits = c(0, 3)) + 
  scale_x_discrete(labels=treatment_names)+
  axis_looks

#save plot
ggsave(p.myc2.Bole,
       height = 3.52, width = 3.55,
       file="./R output/Figures_for_publication/traits_BO_MYC2.svg")
```


### Data analysis 

```{r}

#check distribution
continuous_distribution_plot_table(qpcr.Atha$myc2) #not normal, gamma 
hist(qpcr.Atha$myc2)

#LM
lm.myc2.Atha <- lm(
  myc2 ~ treatment, 
  data = qpcr.Atha)

Anova(lm.myc2.Atha)

simulateResiduals(lm.myc2.Atha, plot = T) #okay 
plot(lm.myc2.Atha)

aov_residuals<-residuals(lm.myc2.Atha) 
shapiro.test(aov_residuals) #okay 

#GLM with Gamma distribution
glm.myc2.Atha <- glm (myc2 ~ treatment,
               family = Gamma (link = "log"),
               data = qpcr.Atha)

simulateResiduals(glm.myc2.Atha, plot = T) #ok
plot(glm.myc2.Atha)

Anova(glm.myc2.Atha)
summary(glm.myc2.Atha)
summ(glm.myc2.Atha, digits = 4)

#Post-hoc
emm.myc2.Atha <- emmeans(ref_grid(lm.myc2.Atha), pairwise ~ treatment, type="response")

plot(emm.myc2.Atha$emmeans, comparisons = TRUE)
pairs(emm.myc2.Atha)
eff_size(emm.myc2.Atha, sigma = sigma(glm.myc2.Atha), edf = 23)

cld.myc2.Atha <- cld(emm.myc2.Atha,
           alpha = 0.05,
           Letters = letters,
           adjust = "tukey")

#Brassica

#check distribution
continuous_distribution_plot_table(qpcr.Bole$myc2) #not normal, gamma or lognormal
hist(qpcr.Bole$myc2)

#LM
lm.myc2.Bole <- lm(
  myc2 ~ treatment, 
  data = qpcr.Bole)

simulateResiduals(lm.myc2.Bole, plot = T) #ok
plot(lm.myc2.Bole)

aov_residuals<-residuals(lm.myc2.Bole) 
shapiro.test(aov_residuals) #not normal 

#GLM with Gamma distribution
glm.myc2.Bole <- glm (myc2 ~ treatment,
               family = Gamma (link = "log"),
               data = qpcr.Bole)

simulateResiduals(glm.myc2.Bole, plot = T) #ok
plot(glm.myc2.Bole)

Anova(glm.myc2.Bole)
summary(glm.myc2.Bole)
summ(glm.myc2.Bole, digits = 4)

#Post-hoc
emm.myc2.Bole <- emmeans(ref_grid(glm.myc2.Bole), pairwise ~ treatment, type="response")

plot(emm.myc2.Bole$emmeans, comparisons = TRUE)
pairs(emm.myc2.Bole)

cld.myc2.Bole <- multcomp::cld(emm.myc2.Bole,
           alpha = 0.05,
           Letters = letters,
           adjust = "tukey")

```

Only for Arabidopsis thaliana there's a significant effect of the treatment in the relative expression levels of MYC2, specifically, plants treated with the highest concentration of MeJA had higher expression of MYC2 compared to controls. 


## 9.10 Cohen's D comparison 

The standardized mean difference (cohen's D) with confidence intervals were calculated for:  

* dry biomass BO
* leaf area BO
* inflorescence weight AT
* Number of siliques AT
* LOX2 expression for both AT and BO 
* MYC2 expression for both AT and BO 

Reference suggested by Pedro: https://nph.onlinelibrary.wiley.com/doi/full/10.1111/nph.15504 

### Cohen's D
```{r}
#making costume function using lsr package to calculate Cohen's D

cohensD_per_trait <- function(df, column, species){
  
  #df= data frame
  #column = column name with trait of interest, no quotes 
  #species = species name to be put in df, no quotes 
  
  trait_name <- deparse(substitute(column)) #gets actual name (" ") of the column to be substituted
  species_name <- deparse(substitute(species)) #idem for species

Ctrl<- df %>%  #gets only values for control and the trait 
       dplyr::filter(treatment == "control") %>%  pull(trait_name) 

MeJA_0.1 <- df %>% 
           dplyr::filter(treatment == "MeJA_0.1") %>% pull(trait_name)

MeJA_1 <- df %>% 
           dplyr::filter(treatment == "MeJA_1.0") %>% pull(trait_name)
  
OS <- df %>% 
           dplyr::filter(treatment == "oral_secretion") %>% pull(trait_name)

d_MeJA0.1 <- cohen.d(MeJA_0.1, Ctrl) #calculates Cohen's D, Ctrl should be 2nd position
d_MeJA1.0 <- cohen.d(MeJA_1, Ctrl)   #so estimate matches the effect: 
d_OS <- cohen.d(OS, Ctrl)            #(smaller or larger than ctrl)

a <- as.data.frame(t(unlist(d_MeJA0.1))) %>% #makes new df with values from Cohen's D estimation
      mutate(treatment="MeJA_0.1")           #and adds a new column with the treatment 

b <- as.data.frame(t(unlist(d_MeJA1.0))) %>% 
      mutate(treatment="MeJA_1.0")  

c <- as.data.frame(t(unlist(d_OS))) %>% 
      mutate(treatment="oral_secretion") 

df.cohensD <- rbind(a,b,c)%>%             #merge 3 df's
              mutate(trait=trait_name)%>% #makes a new column with the trait name 
              mutate(species=species_name)%>% #idem for species 
              mutate(eff_size=                #Adds a new column replacing numeric "magnitude" 
                       case_when(             #values (1-4) coming from cohen.d function 
                         magnitude == 1 ~ "negligible", #with the category written in text 
                         magnitude == 2 ~ "small",
                         magnitude == 3 ~ "medium",
                         magnitude == 4 ~ "large"))%>%
             relocate(trait, species, treatment, eff_size)


return(df.cohensD) #returns full df 
}  

#calculate cohen's D for each trait 
d_drybiomass <- cohensD_per_trait(Bole, ldry_weight, Brassica_oleraceae)
d_leaf_area <- cohensD_per_trait(Bole, leaf_area, Brassica_oleraceae)
d_lox2.bo <- cohensD_per_trait(qpcr.Bole, lox2, Brassica_oleraceae)
d_myc2.bo <- cohensD_per_trait(qpcr.Bole, myc2, Brassica_oleraceae)
d_number_fruits <- cohensD_per_trait(Atha, number_fruits, Arabidopsis_thaliana)
d_inflorescence <- cohensD_per_trait(Atha, inflorescence, Arabidopsis_thaliana)
d_lox2.at <- cohensD_per_trait(qpcr.Atha, lox2, Arabidopsis_thaliana)
d_myc2.at <- cohensD_per_trait(qpcr.Atha, myc2, Arabidopsis_thaliana)

#merge df's
d_traits <- rbind(d_drybiomass, d_leaf_area, d_lox2.bo, d_myc2.bo, d_number_fruits,
                  d_inflorescence, d_lox2.at, d_myc2.at)

#changing to factor and numeric 
d_traits[c(1:4)] <- lapply(d_traits[c(1:4)], as.factor)
d_traits[c(7:13)] <- lapply(d_traits[c(7:13)], as.numeric)

#order traits
d_traits$trait = factor(d_traits$trait,
                         level=c("myc2", "lox2", "number_fruits", "inflorescence", "leaf_area", "ldry_weight"))

```

### Plots  

Followed this: https://stackoverflow.com/questions/58657802/forest-plot-with-subgroups-in-ggplot2 

```{r}
#settings for plots 
color_treatments_noControl <- c("#DEA868", "#C7522B", "#9BBAA0")
names_treatments_noControl <- c(c("MeJA_0.1" = "0.1 mM \n MeJA", "MeJA_1.0" = "1 mM \n MeJA", "oral_secretion"="P. brassicae \n OS"))
trait_names <- c("myc2" = "relative expression \n of MYC2", "lox2" = "relative expression \n of LOX2",
                      "number_fruits" = "Number of siliques \n per plant", "inflorescence"="Dry weight \n inflorescence", "ldry_weight"="Dry weight \n leaves", "leaf_area"="Leaf area")


#Arabidopsis
p.cohensD.Atha <- ggplot(data = subset(d_traits, species == "Arabidopsis_thaliana"), mapping = aes(x = trait, y = estimate,
                    ymin=conf.int.lower, ymax=conf.int.upper,colour= treatment, fill=treatment))+
  geom_linerange(size=5, position=position_dodge(width = 0.5)) +
  geom_hline(aes(yintercept=0), lty=2) +
    geom_hline(aes(yintercept=0.8), lty=2, colour="red") +
  geom_hline(aes(yintercept=-0.8), lty=2, colour="red") +
  geom_point(size=3, shape=20, colour="black", stroke = 0.5,
             position=position_dodge(width = 0.5)) +
  scale_fill_manual(values=color_treatments_noControl)+
  scale_color_manual(values=color_treatments_noControl)+
  scale_x_discrete(name="", labels=trait_names) +
  labs(title = "Arabidopsis thaliana")+
  scale_y_continuous(name="Cohen's D estimate with 95% confidence interval", limits = c(-10.15,5)) +
  coord_flip() +
  theme_minimal()+
  axis_looks+
  theme(axis.text.y = element_text(colour = "black", size = 12,
                                               face = "bold", angle=0, hjust=0.5))+
  theme(legend.position="top")

#Brassica
p.cohensD.Bole <- ggplot(data = subset(d_traits, species == "Brassica_oleraceae"), mapping = aes(x = trait, y = estimate, ymin=conf.int.lower, ymax=conf.int.upper,colour= treatment, fill=treatment))+
  geom_linerange(size=5, position=position_dodge(width = 0.5)) +
  geom_hline(aes(yintercept=0), lty=2) +
    geom_hline(aes(yintercept=0.8), lty=2, colour="red") +
  geom_hline(aes(yintercept=-0.8), lty=2, colour="red") +
  geom_point(size=3, shape=20, colour="black", stroke = 0.5,
             position=position_dodge(width = 0.5)) +
  scale_fill_manual(values=color_treatments_noControl)+
  scale_color_manual(values=color_treatments_noControl)+
  scale_x_discrete(name="", labels=trait_names) +
  labs(title = "Brassica oleracea")+
  scale_y_continuous(name="Cohen's D estimate with 95% confidence interval", limits = c(-5,5)) +
  coord_flip() +
  theme_minimal()+
  axis_looks+
  theme(axis.text.y = element_text(colour = "black", size = 12,
                                               face = "bold", angle=0, hjust=0.5))+
  theme(legend.position="top")

#save plots
ggsave(p.cohensD.Atha,
       height = 4.81, width = 7.04,
       file="./R output/Figures_for_publication/traits_AT_cohensD.svg")

ggsave(p.cohensD.Bole,
       height = 4.81, width = 7.04,
       file="./R output/Figures_for_publication/traits_BO_cohensD.svg")

```

---Cancelled
Now, make a table to be placed next to the plot. 
From: https://rpubs.com/mbounthavong/forest_plots_r

```{r}

## Create the table-base pallete
table_base <- ggplot(data = subset(d_traits, species == "Arabidopsis_thaliana"), aes(y=trait, fill=treatment)) +
  ylab(NULL) + xlab("  ") + 
  theme(plot.title = element_text(hjust = 0.5, size=12), 
        axis.text.x = element_text(color="white", hjust = -3, size = 25), ## This is used to help with alignment
        axis.line = element_blank(),
        axis.text.y = element_blank(), 
        axis.ticks = element_blank(),
        axis.title.y = element_blank(), 
        legend.position = "none",
        panel.background = element_blank(), 
        panel.border = element_blank(), 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), 
        plot.background = element_blank())

## OR point estimate table
tab1 <- table_base + 
  labs(title = "space") +
  geom_text(aes(y = rev(trait), x = 1, label = sprintf("%0.1f", round(estimate, digits = 2))), size = 4) + ## decimal places
  ggtitle("Cohen's D")

## 95% CI table
tab2 <- table_base +
  geom_text(aes(y = rev(trait), x = 1, label = eff_size), size = 4) + 
  ggtitle("Effect size")


## Merge tables with plot
lay <-  matrix(c(1,1,1,1,1,1,1,1,1,1,2,3,3), nrow = 1)
grid.arrange(p.cohensD.Atha, tab1, tab2, layout_matrix = lay)


```

It's difficult to align with groups per treatment, I will stop and add it on inkscape, perhaps it's better to add an '**' when effects are large. 
---Cancelled

## 10.X Cohen's D without geneExp

Link for fixing distance between bars:
https://stackoverflow.com/questions/51892875/how-to-increase-the-space-between-grouped-bars-in-ggplot2 

###Dataframe

```{r}

###dataframe
###dataframe
library(ggplot2)
library(tidyr)
library(grid)
library(gridExtra)
library(forcats)


forestdf <- structure(list(labels = structure(1:36, .Label = c("Age*", "Sex – male vs. female", 
                                                               "Body-mass index*,1 ", "Systolic blood pressure*", "Race - vs. white", 
                                                               "Asian", "Black", "Townsend deprivation index", "Social habit", 
                                                               "Smoking - vs. never", "Previous", "Current", "Alcohol use - vs. never", 
                                                               "Once or twice a week", "Three or four times a week", "Daily or almost daily", 
                                                               "Comorbidity", "Cancer", "Diabetes", "Chronic obstructive pulmonary disease2", 
                                                               "Asthma", "Ischemic heart disease3", "Hypothyroidism", "Hypercholesterolemia", 
                                                               "Allergic rhinitis", "Depression", "Serology", "White blood cell count", 
                                                               "Red blood cell count", "Hemoglobin concentration", "Mean corpuscular volume", 
                                                               "Mean corpuscular hemoglobin concentration", "Platelet count", 
                                                               "Lymphocyte count", "Monocyte count", "Neutrophil count"), class = "factor"), 
                           rr = c(1.18, 1.45, 1.76, 0.98, NA, 2.16, 2.65, 1.09, NA, 
                                  NA, 1.35, 1.15, NA, 0.73, 0.63, 0.63, NA, 1.23, 1.34, 1.51, 
                                  1.12, 1.46, 0.96, 1.1, 1.18, 1.38, NA, 1.03, 0.87, 0.93, 
                                  1, 0.94, 1, 1.03, 1.17, 1.06), rrhigh = c(1.08, 1.28, 1.57, 
                                                                            0.95, NA, 1.63, 2.03, 1.07, NA, NA, 1.18, 0.94, NA, 0.58, 
                                                                            0.49, 0.5, NA, 0.99, 1.08, 1.09, 0.93, 1.15, 0.71, 0.92, 
                                                                            0.91, 1.1, NA, 1.02, 0.73, 0.87, 0.99, 0.88, 1, 1.01, 1.03, 
                                                                            1.01), rrlow = c(1.28, 1.64, 1.97, 1.02, NA, 2.86, 3.44, 
                                                                                             1.11, NA, NA, 1.55, 1.42, NA, 0.9, 0.79, 0.81, NA, 1.53, 
                                                                                             1.66, 2.09, 1.34, 1.85, 1.3, 1.31, 1.52, 1.74, NA, 1.04, 
                                                                                             1.03, 0.98, 1.01, 1.01, 1, 1.05, 1.32, 1.1)), class = "data.frame", row.names = c(NA, -36L))


forestdf$labels <- factor(forestdf$labels,levels = forestdf$labels)
levels(forestdf$labels)                                                                                                                                                                                                           #1.52, 1.74, NA, 1.04, 1.03, 0.98, 1.01, 1.01, 1, 1.05, 1.32, 

#forestplot
p <- ggplot(forestdf, aes(x=rr, y=labels, xmin=rrlow, xmax=rrhigh))+
  geom_pointrange(shape=22, fill="black")+
  geom_vline(xintercept = 1, linetype=3)+
  xlab("Variable")+ylab("Adjusted Relative Risk with 95% Confidence Interval")+theme_classic()+scale_y_discrete(limits = rev(labels))+
  scale_x_log10(limits = c(0.25, 4), breaks = c(0.25, 0.5, 1, 2, 4), labels=c("0.25", "0.5", "1", "2", "4"), expand = c(0,0))
p



#dataframe for table
fplottable <- structure(list(labels = structure(c(1L, 30L, 7L, 33L, 27L, 4L, 
6L, 35L, 32L, 31L, 26L, 11L, 2L, 24L, 34L, 12L, 10L, 8L, 14L, 
9L, 5L, 18L, 17L, 16L, 3L, 13L, 29L, 36L, 28L, 15L, 21L, 20L, 
25L, 19L, 22L, 23L), .Label = c("Age*", "Alcohol use - vs. never", 
"Allergic rhinitis", "Asian", "Asthma", "Black", "Body-mass index*,1 ", 
"Cancer", "Chronic obstructive pulmonary disease2", "Comorbidity", 
"Current", "Daily or almost daily", "Depression", "Diabetes", 
"Hemoglobin concentration", "Hypercholesterolemia", "Hypothyroidism", 
"Ischemic heart disease3", "Lymphocyte count", "Mean corpuscular hemoglobin concentration", 
"Mean corpuscular volume", "Monocyte count", "Neutrophil count", 
"Once or twice a week", "Platelet count", "Previous", "Race - vs. white", 
"Red blood cell count", "Serology", "Sex – male vs. female", 
"Smoking - vs. never", "Social habit", "Systolic blood pressure*", 
"Three or four times a week", "Townsend deprivation index", "White blood cell count"
), class = "factor"), No..of.Events = c(1073L, 581L, 1061L, 1031L, 
NA, 57L, 68L, 1072L, NA, NA, 442L, 117L, NA, 262L, 191L, 172L, 
NA, 96L, 107L, 41L, 146L, 86L, 52L, 170L, 66L, 84L, NA, 1009L, 
1009L, 1009L, 1009L, 1009L, 1009L, 1005L, 1005L, 1005L), ARR..95..CI. = c("1.18 (1.08-1.28)", 
"1.45 (1.28-1.64)", "1.76 (1.57-1.97)", "0.98 (0.95-1.02)", "", 
"2.16 (1.63-2.86)", "2.65 (2.03-3.44)", "1.09 (1.07-1.11)", "", 
"", "1.35 (1.18-1.55)", "1.15 (0.94-1.42)", "", "0.73 (0.58-0.90)", 
"0.63 (0.49-0.79)", "0.63 (0.50-0.81)", "", "1.23 (0.99-1.53)", 
"1.34 (1.08-1.66)", "1.51 (1.09-2.09)", "1.12 (0.93-1.34)", "1.46 (1.15-1.85)", 
"0.96 (0.71-1.30)", "1.10 (0.92-1.31)", "1.18 (0.91-1.52)", "1.38 (1.10-1.74)", 
"", "1.03 (1.02-1.04)", "0.87 (0.73-1.03)", "0.93 (0.87-0.98)", 
"1.00 (0.99-1.01)", "0.94 (0.88-1.01)", "1.00 (1.00-1.00)", "1.03 (1.01-1.05)", 
"1.17 (1.03-1.32)", "1.06 (1.01-1.10)")), class = "data.frame", row.names = c(NA, 
-36L))

###NOT WORKING CODE THAT TRIES TO MAKE TABLE LEFT OF FOREST PLOT
data_table <- geom_text(data=fplottable,aes(y=labels)) +
  geom_text(label=eventnum) +
  geom_text(label=arr)
data_table

grid.arrange(data_table,p, ncol=2)
######



forestdf$colour <- rep(c("white", "gray95"), 18)

p <- ggplot(forestdf, aes(x = rr, y = labels, xmin = rrlow, xmax = rrhigh)) +
  geom_hline(aes(yintercept = labels, colour = colour), size = 7) + 
  geom_pointrange(shape = 22, fill = "black") +
  geom_vline(xintercept = 1, linetype = 3) +
  xlab("Variable") +
  ylab("Adjusted Relative Risk with 95% Confidence Interval") +
  theme_classic() +
  scale_colour_identity() +
  scale_y_discrete(limits = rev(forestdf$labels)) +
  scale_x_log10(limits = c(0.25, 4), 
                breaks = c(0.25, 0.5, 1, 2, 4), 
                labels = c("0.25", "0.5", "1", "2", "4"), expand = c(0,0)) +
  theme(axis.text.y = element_blank(), axis.title.y = element_blank())



names(fplottable) <- c("labels", "eventnum", "arr")
fplottable$labels <- factor(fplottable$labels, rev(levels(forestdf$labels)))
fplottable$colour <- rep(c("white", "gray95"), 18)

data_table <- ggplot(data = fplottable, aes(y = labels)) +
  geom_hline(aes(yintercept = labels, colour = colour), size = 7) +
  geom_text(aes(x = 0, label = labels), hjust = 0) +
  geom_text(aes(x = 5, label = eventnum)) +
  geom_text(aes(x = 7, label = arr), hjust = 1) +
  scale_colour_identity() +
  theme_void() + 
  theme(plot.margin = margin(5, 0, 35, 0))

grid.arrange(data_table,p, ncol = 2)





```






###Plots
```{r}

#filter traits df 
d_traits.s <- rbind(d_drybiomass, d_leaf_area, d_number_fruits, d_inflorescence)

#changing to factor and numeric 
d_traits.s[c(1:4)] <- lapply(d_traits.s[c(1:4)], as.factor)
d_traits.s[c(7:13)] <- lapply(d_traits.s[c(7:13)], as.numeric)

#order
d_traits.s$trait = factor(d_traits.s$trait,
                         level=c("leaf_area","ldry_weight","number_fruits","inflorescence"))

#settings for plots 
color_treatments_noControl <- c("#DEA868", "#C7522B", "#9BBAA0")
names_treatments_noControl <- c(c("MeJA_0.1" = "0.1 mM \n MeJA", "MeJA_1.0" = "1 mM \n MeJA", "oral_secretion"="P. brassicae \n OS"))
trait_names <- c("number_fruits" = "Number of siliques \n per plant", "inflorescence"="Dry weight \n inflorescence", "ldry_weight"="Dry weight \n leaves", "leaf_area"="Leaf area")

#Both species
p.cohensD.2 <- ggplot(data = d_traits.s, mapping = aes(x = trait, y = estimate, colour= treatment, fill=treatment))+
  #geom_linerange(size=5, position=position_dodge(width = 0.5)) +
  geom_hline(aes(yintercept=0), lty=1, size=0.8) +
    geom_hline(aes(yintercept=0.8), lty=2, size=0.8, colour="gray") +
  geom_hline(aes(yintercept=-0.8), lty=2, size=0.8, colour="gray") +
  geom_col(colour="black", width=0.7, position = position_dodge(0.7))+
  scale_fill_manual(values=color_treatments_noControl)+
  scale_color_manual(values=color_treatments_noControl)+
  scale_x_discrete(name="", labels=trait_names) +
  labs(title = "AT_BO")+
  scale_y_continuous(name="Cohen's D estimate", limits = c(-2.5,1.5)) +
  coord_flip() +
  theme_minimal()+
  axis_looks+
  theme(axis.text.y = element_text(colour = "black", size = 12,
                     face = "bold", angle=0, hjust=0.5))+
  theme(legend.position="top")


#Arabidopsis
p.cohensD.Atha2 <- ggplot(data = subset(d_traits.s, species == "Arabidopsis_thaliana"), mapping = aes(x = trait, y = estimate, colour= treatment, fill=treatment))+
  #geom_linerange(size=5, position=position_dodge(width = 0.5)) +
  geom_hline(aes(yintercept=0), lty=2) +
    geom_hline(aes(yintercept=0.8), lty=2, colour="red") +
  geom_hline(aes(yintercept=-0.8), lty=2, colour="red") +
  geom_bar(stat='identity', position = "dodge2")+
  scale_fill_manual(values=color_treatments_noControl)+
  scale_color_manual(values=color_treatments_noControl)+
  scale_x_discrete(name="", labels=trait_names) +
  labs(title = "Arabidopsis thaliana")+
  scale_y_continuous(name="Cohen's D estimate", limits = c(-3,3)) +
  coord_flip() +
  theme_minimal()+
  axis_looks+
  theme(axis.text.y = element_text(colour = "black", size = 12,
                                               face = "bold", angle=0, hjust=0.5))+
  theme(legend.position="top")

#Brassica
p.cohensD.Bole2 <- ggplot(data = subset(d_traits.s, species == "Brassica_oleraceae"), mapping = aes(x = trait, y = estimate, colour= treatment, fill=treatment))+
  #geom_linerange(size=5, position=position_dodge(width = 0.5)) +
  geom_hline(aes(yintercept=0), lty=2) +
    geom_hline(aes(yintercept=0.8), lty=2, colour="red") +
  geom_hline(aes(yintercept=-0.8), lty=2, colour="red") +
  geom_bar(stat='identity', position = "dodge2")+
  scale_fill_manual(values=color_treatments_noControl)+
  scale_color_manual(values=color_treatments_noControl)+
  scale_x_discrete(name="", labels=trait_names) +
  labs(title = "Brassica oleracea")+
  scale_y_continuous(name="Cohen's D estimate", limits = c(-3,3)) +
  coord_flip() +
  theme_minimal()+
  axis_looks+
  theme(axis.text.y = element_text(colour = "black", size = 12,
                                   face = "bold", angle=0, hjust=0.5))+
  theme(legend.position="top")


#save plot
ggsave(p.cohensD.2,
       height = 4.81, width = 7.04,
       file="./R output/Figures_for_publication/traits_AT_BO_cohensD.svg")


```
Missing:

*Make DF with 
 1. Confidence intervals 
 2. Type of test 
 3. P values 
 
*Add pairwise contrasts (--> How?) 



